File: compare/hard/sqrt_sum.cpp

#include <chrono>
#include <cmath>
#include <iostream>

int main() {
  const long N = 100000000;
  double sum = 0.0;

  auto start = std::chrono::high_resolution_clock::now();

  std::cout << "Starting C++ calculation..." << std::endl;
  for (long i = 1; i <= N; ++i) {
    sum += std::sqrt(i);
    if (i % (N / 10) == 0) {
      std::cout << "C++ progress: " << (i * 100 / N) << "%" << std::endl;
    }
  }

  auto end = std::chrono::high_resolution_clock::now();
  std::chrono::duration<double> duration = end - start;

  std::cout << "C++ Sum: " << sum << std::endl;
  std::cout << "C++ Calculation Time: " << duration.count() << " seconds"
            << std::endl;
  return 0;
}

File: compare/hard/sqrt_sum.f90
PROGRAM sqrt_sum
  IMPLICIT NONE
  INTEGER, PARAMETER :: N = 100000000
  INTEGER :: i
  REAL(8) :: sum
  REAL(8) :: start_time, end_time

  sum = 0.0D0
  CALL CPU_TIME(start_time)
  PRINT *, "Starting Fortran calculation..."
  
  DO i = 1, N
    sum = sum + SQRT(REAL(i, 8))
    IF (MOD(i, N / 10) == 0) THEN
      PRINT *, "Fortran progress: ", i * 100 / N, "%"
    END IF
  END DO

  CALL CPU_TIME(end_time)

  PRINT *, "Fortran Sum: ", sum
  PRINT *, "Fortran Calculation Time: ", end_time - start_time, " seconds"
END PROGRAM sqrt_sum
